process_tree/process_tree.py

```
@@ -0,0 +1,558 @@
+        except subprocess.CalledProcessError as e:
+            print(f"Error processing {recording_file}: {e}", file=sys.stderr)
+            if e.stderr:
+                print(f"  stderr: {e.stderr.strip()}", file=sys.stderr)
+            return None, None, 0.0
+        except (json.JSONDecodeError, KeyError, ValueError) as e:
+            print(f"Error processing {recording_file}: {e}", file=sys.stderr)
+            return None, None, 0.0
+
+    def _run_recording_json(self, recording_file: Path, section: str) -> dict:
+        """Run undo recording-json and return parsed JSON data."""
+        cmd = ["undo", "recording-json", "-s", section, str(recording_file)]
+        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
+        return json.loads(result.stdout)
+
+    def _get_process_ids(self, recording_file: Path) -> Tuple[Optional[int], Optional[int]]:
+        """Get PID and PPID from recording file."""
+        data = self._run_recording_json(recording_file, "d")
+        pid = data["debuggee"]["state_load_rchild_pid"]
+        ppid = data["debuggee"]["rchild_ppid"]
+        return pid, ppid
+
+    def _get_start_time(self, recording_file: Path) -> float:
+        """Get start time from recording file header."""
+        data = self._run_recording_json(recording_file, "h")
+        utc_start = data["header"]["utc_start"]
+        utc_start_ns = data["header"]["utc_start_ns"]
+        try:
+            return float(utc_start) + float(utc_start_ns) / 1_000_000_000
+        except (ValueError, TypeError) as e:
+            raise ValueError(f"Invalid timestamp data in recording header: {e}") from e
+
+
+class ProcessTree:
+    """Represents and manages a tree of processes."""
+
+    def __init__(self):
+        self.processes: Dict[int, Process] = {}
+        self.root: Optional[Process] = None
+
+    def add_process(self, process: Process) -> None:
+        """Add a process to the tree."""
+        self.processes[process.pid] = process
```

You could end up with multiple recordings sharing the same PID if you save multiple times. Maybe we could report an error if that happens. Or at least, add a comment.